File: Python/textbook_track/chapter05/ba5b.py
# Find the Length of a Longest Path in a Manhattan-like Grid
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def read_weight_matrix(reader, nr_rows):
    matrix = []
    for _ in range(nr_rows):
        row = convert_to_intlist(next(reader))
        matrix.append(row)
    return matrix


def calc_longest_path(n, m, down_weights, right_weights):
    longest_path = [[0] * (m + 1) for _ in range(n+1)]
    for ix in range(1, n+1):
        longest_path[ix][0] = longest_path[ix-1][0] + down_weights[ix-1][0]
    for jy in range(1, m+1):
        longest_path[0][jy] = longest_path[0][jy-1] + right_weights[0][jy-1]

    for ix in range(1, n+1):
        for jy in range(1, m+1):
            path_down = longest_path[ix-1][jy] + down_weights[ix-1][jy]
            path_right = longest_path[ix][jy-1] + right_weights[ix][jy-1]
            longest_path[ix][jy] = max(path_down, path_right)
    return longest_path[n][m]


def main():
    reader = sys.stdin
    n, m = convert_to_intlist(next(reader))
    down = read_weight_matrix(reader, n)
    _ = next(reader)
    right = read_weight_matrix(reader, n+1)
    result = calc_longest_path(n, m, down, right)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter05/BA5B.scala
package TextbookTrack.Chapter05

object BA5B {
  type WeightMatrix = Vector[Vector[Int]]

  private def convertToIntVector(line: String): Vector[Int] = line.split(" ").map(_.toInt).toVector

  private def readParameters(line: String): (Int, Int) = convertToIntVector(line) match {
    case Vector(n, m) => (n, m)
    case _ => throw new Exception("Unexpected number of parameters.")
  }

  private def readWeightMatrix(lines: Iterator[String]): WeightMatrix = lines.map(convertToIntVector).toVector

  def calcLongestPath(n: Int, m: Int, downWeights: WeightMatrix, rightWeights: WeightMatrix): Int = {
    val longestPath: Array[Array[Int]] = Array.fill(n + 1, m + 1)(0)
    (1 to n).foreach(ix => longestPath(ix)(0) = longestPath(ix - 1)(0) + downWeights(ix - 1)(0))
    (1 to m).foreach(jy => longestPath(0)(jy) = longestPath(0)(jy - 1) + rightWeights(0)(jy - 1))

    for {
      ix <- 1 to n
      jy <- 1 to m
    } {
      val pathDown: Int = longestPath(ix - 1)(jy) + downWeights(ix - 1)(jy)
      val pathRight: Int = longestPath(ix)(jy - 1) + rightWeights(ix)(jy - 1)
      longestPath(ix)(jy) = math.max(pathDown, pathRight)
    }
    longestPath(n)(m)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (n, m): (Int, Int) = readParameters(reader.next())
    val down: WeightMatrix = readWeightMatrix(reader.take(n))
    val _: String = reader.next()
    val right: WeightMatrix = readWeightMatrix(reader.take(n + 1))
    val result: Int = calcLongestPath(n, m, down, right)
    println(result)
  }
}
